src/programacion/empresaprogramacion/Empresa.java

```diff
@@ -1,54 +1,216 @@
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
+import java.util.TreeMap;
 
 public class Empresa {
-    
-    public String nombre;
-    public Map<Integer, Empleado> mapaEmpleados;
 
-    public Empresa(String nombre) {
-        this.nombre = nombre;
-        //TODO this.empleados = new ?<>();
-    }
+  public String nombre;
+  public Map<String, Empleado> mapaEmpleados;
+  static final String CSV_FILENAME_PROGRAMADORES = "programadores.csv";
+  static final String CSV_FILENAME_GERENTES = "gerentes.csv";
 
-    public void addEmpleado(Empleado empleado) {
-        // TODO  (no permitir meter 2 empleados con el mismo "idEmpleado" y dado el caso lanzar 
-        //exception)
+  public Empresa(String nombre) {
+    this.nombre = nombre;
+    mapaEmpleados = new HashMap<>();
+  }
+
+  public void addEmpleado(Empleado empleado) throws ParametroInvalidoException {
+    if (mapaEmpleados.containsKey(empleado.DNI)) {
+      if (empleado.tipoEmpleado == TipoEmpleado.PROGRAMADOR) {
+        throw new ParametroInvalidoException("Ya existe un Programador con ese ID de empleado.\n");
+      } else {
+        throw new ParametroInvalidoException("Ya existe un Gerente con ese ID de empleado.\n");
+      }
+    } else {
+      mapaEmpleados.put(empleado.DNI, empleado);
     }
+  }
+
+  public void guardarEnCSV(String archivoGuardar) throws RutaArchivoInvalida {
+    List<Programador> programadores = new ArrayList<>();
+    List<Gerente> gerentes = new ArrayList<>();
 
-    public void guardarEnCSV(String archivoGuardar) {
-        //TODO 
+    for (Empleado empleado : mapaEmpleados.values()) {
+      if(empleado instanceof Programador){
+        programadores.add((Programador) empleado);
+      } else {
+        gerentes.add((Gerente) empleado);
+      }
     }
 
-    public void cargarDesdeCSV(String archivoCargar) {
-        //TODO
+    Iterator<Programador> iteradorProgramadores = programadores.iterator();
+    Iterator<Gerente> iteradorGerentes = gerentes.iterator();
 
+    try{
+      File archivo = new File(archivoGuardar);
+      if (archivoGuardar.equalsIgnoreCase(CSV_FILENAME_PROGRAMADORES)){
+        if (comprobarArchivo(archivo)){
+          if (archivo.createNewFile()){
+            System.out.println("El archivo se ha creado correctamente.");
+          } else {
+            throw new RutaArchivoInvalida("El archivo no se ha podido crear.");
+          }
+        }
+        try (BufferedWriter bw = new BufferedWriter(new FileWriter(archivo))){
+          bw.write(Programador.CSV_HEADER);
+          bw.newLine();
+          while(iteradorProgramadores.hasNext()){
+            bw.write(iteradorProgramadores.next().toCsvLine());
+            bw.newLine();
+          }
+          bw.flush();
```

Comment:
No llaméis a flush(), no hace falta, cuando se cierra el fichero ya se hace el flush() automáticamente. No es buena práctica. 
Si llamamos a flush podemos impedir optimizaciones del sistema de ficheros o del sistema operativo.

Reply:
La llamada a flush() era porque no actualizaba el archivo durante la ejecución del programa sino al finalizarla. Por si se carga un archivo, se escribe otro y se vuelve a cargar, a mi al menos no me lo hacía. Si hubiera un finally con close() o una llamada a close() a mano, si que lo hacía, pero el try con resources parece cerrarlo (al menos en este programa tan pequeño) al finalizar la ejecución, y es al llamar a close() cuando escribe el fichero (porque close() implica flush()). ¿Entiendo que eso te da igual y que lo sobreentiendes, no?